dashboard/app/controllers/api_controller.rb

```diff
@@ -546,6 +547,120 @@ def assessments_structure
     render json: assessments
   end
 
+  # Return a hash by student_id for all students in a section. The value is a hash by
+  # script_level_id, with information on the student responses for that assessment.
+  # Currently, very similar logic to section_assessments, but will change as we refactor. The format
+  # of the results is different enough that I'm duplicating some code it for now.
+  # TODO(caleybrock): currently only used in internal experiment, must add controller tests.
+  def assessments_responses
+    section = load_section
+    script = load_script(section)
+
+    responses_by_student = {}
+
+    level_group_script_levels = script.script_levels.includes(:levels).where('levels.type' => 'LevelGroup')
+
+    section.students.each do |student|
+      # Initialize student hash
+      student_hash = {
+        student_name: student.name
+      }
+      responses_by_level_group = {}
+
+      level_group_script_levels.each do |script_level|
```

Comment:
This method is huge!  I'd take each of these subsections where you're building up a hash as a potential method extraction point.